### PR TITLE
Deprecate `amp_level` from `Trainer`

### LINKED ISSUE #12655: [RFC] Deprecate `amp_level` in favour of `ApexMixedPrecisionPlugin(amp_level=...)`
## Proposed refactor

Deprecate `amp_level` and let users choose a different `amp_level` by passing `Trainer(..., plugins=[ApexMixedPrecisionPlugin(amp_level=...)]`.

Previous discussion: https://github.com/PyTorchLightning/pytorch-lightning/issues/9956#issuecomment-1091645295

### Motivation

Currently, `amp_level` is used only when `precision=16` and `amp_backend='apex'` i.e it's only relevant for only one type of configuration and is generally not required.

Users can easily choose it by setting `Trainer(..., plugins=[ApexMixedPrecisionPlugin(amp_level=...)]` which IMO makes more sense since it now reflects that amp_level is only relevant for `ApexMixedPrecisionPlugin`.

### Pitch

Deprecate in v1.6 and remove in v1.8.

### Additional context

______________________________________________________________________

#### If you enjoy Lightning, check out our other projects! ⚡

- [**Metrics**](https://github.com/PyTorchLightning/metrics): Machine learning metrics for distributed, scalable PyTorch applications.

- [**Lite**](https://pytorch-lightning.readthedocs.io/en/latest/starter/lightning_lite.html): enables pure PyTorch users to scale their existing code on any kind of device while retaining full control over their own loops and optimization logic.

- [**Flash**](https://github.com/PyTorchLightning/lightning-flash): The fastest way to get a Lightning baseline! A collection of tasks for fast prototyping, baselining, fine-tuning, and solving problems with deep learning.

- [**Bolts**](https://github.com/PyTorchLightning/lightning-bolts): Pretrained SOTA Deep Learning models, callbacks, and more for research and production with PyTorch Lightning and PyTorch.

- [**Lightning Transformers**](https://github.com/PyTorchLightning/lightning-transformers): Flexible interface for high-performance research using SOTA Transformers leveraging Pytorch Lightning, Transformers, and Hydra.

cc @justusschock @awaelchli @rohitgr7 @tchaton @kaushikb11 @borda

### PR DESCRIPTION
**Pull request by rohitgr7**
## What does this PR do?

Fixes #12655

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

cc @tchaton @rohitgr7 @carmocca @justusschock @awaelchli @akihironitta @kaushikb11 @borda

### CODE REVIEW

File: src/pytorch_lightning/trainer/connectors/accelerator_connector.py [hunk 1]
```diff
@@ -188,6 +188,12 @@ def __init__(
         self._amp_level_flag: Optional[str] = amp_level
         self._auto_select_gpus: bool = auto_select_gpus
 
+        if amp_level is not None:
+            rank_zero_deprecation(
+                "Setting `amp_level` inside the `Trainer` is deprecated in v1.8.0 and will be removed"
+                " in v2.0.0. Please set it inside the specific precision plugin and pass it to the `Trainer`."
```
Comment (carmocca) [inline src/pytorch_lightning/trainer/connectors/accelerator_connector.py:194]:
```suggestion
                " in v1.10.0. Please set it inside the specific precision plugin and pass it to the `Trainer`."
```

File: tests/tests_pytorch/deprecated_api/test_remove_2-0.py [hunk 2]
```diff
@@ -88,3 +88,8 @@ def test_v2_0_resume_from_checkpoint_trainer_constructor(tmpdir):
         trainer = Trainer(resume_from_checkpoint="trainer_arg_path")
     with pytest.raises(FileNotFoundError, match="Checkpoint at fit_arg_ckpt_path not found. Aborting training."):
         trainer.fit(model, ckpt_path="fit_arg_ckpt_path")
+
+
+def test_deprecated_amp_level():
```
Comment (carmocca) [inline tests/tests_pytorch/deprecated_api/test_remove_2-0.py:93]: There is no date for 2.0 yet. So it's safer to set the removal with `1.10`

**Review by carmocca — APPROVED** [review]
LGTM other than the removal version

File: tests/tests_pytorch/deprecated_api/test_remove_2-0.py (hunk 2 above)
Comment (rohitgr7) [reply tests/tests_pytorch/deprecated_api/test_remove_2-0.py]: done!

File: src/pytorch_lightning/trainer/trainer.py [hunk 3]
```diff
@@ -209,6 +209,10 @@ def __init__(
             amp_level: The optimization level to use (O1, O2, etc...). By default it will be set to "O2"
                 if ``amp_backend`` is set to "apex".
 
+                .. deprecated:: v1.5
```
Comment (awaelchli) [inline src/pytorch_lightning/trainer/trainer.py:212]:
```suggestion
                .. deprecated:: v1.8
```

**Review by awaelchli — APPROVED** [review]

**Review by akihironitta — APPROVED** [review]
LGTM!
